runtime/pubsub/internal/encorecloud/push_handler.go

```diff
@@ -0,0 +1,187 @@
+type pushPayload struct {
+	Data            []byte            `json:"data"`
+	Attributes      map[string]string `json:"attributes"`
+	MessageID       string            `json:"messageId"`
+	PublishTime     time.Time         `json:"publishTime"`
+	DeliveryAttempt int               `json:"deliveryAttempt"`
+}
+
+// registerPushEndpoint registers a push endpoint for a subscription from Encore Cloud
+//
+// Encore Cloud will send a POST request to the endpoint with a JSON encoded [pushPayload] as the body.
+// The request will be signed with the latest Encore Cloud auth key for this application.
+//
+// Once the request is received and verified, the user's subscription function will be called with the decoded
+// payload, while simultaneously an event stream will be sent back to Encore Cloud to indicate that the request
+// is being processed, with keepalive messages being sent every 5 seconds.
+//
+// If the subscription function returns an error, the event stream will be closed with the error message.
+// If the subscription function returns successfully, the event stream will be closed with a success message.
+//
+// The Encore Cloud server will wait for a valid end response from the event stream before closing the connection and
+// acknowledging the message with the underlying message broker.
+//
+// If the event stream is closed without a valid end response, the message will be nacked and retried by Encore Cloud.
+//
+// If the request is closed by Encore Cloud while a subscription function is still running, the context of the function
+// will be cancelled, as this means Encore Cloud has failed to receive a keepalive message from the event stream and has
+// assumed the request has failed.
+//
+// The events on the stream will be one of these types:
+// - "keepalive"
+// - "ack"
+// - "nack"
+func (mgr *Manager) registerPushEndpoint(logger *zerolog.Logger, subscriptionConfig *config.PubsubSubscription, f types.RawSubscriptionCallback) {
+	mgr.pushRegistry.RegisterPushSubscriptionHandler(
+		types.SubscriptionID(subscriptionConfig.ID),
+		func(w http.ResponseWriter, req *http.Request) {
+			payload, err := mgr.decodeAndVerifyPushPayload(req, subscriptionConfig.ID)
+			if err != nil {
+				logger.Err(err).Msg("error while verifying PubSub subscription message")
+				errs.HTTPError(w, err)
+				return
+			}
+
+			flusher, ok := w.(http.Flusher)
+			if !ok {
+				err := errs.B().Code(errs.Internal).Msg("unable to cast http.ResponseWriter to http.Flusher").Err()
+				logger.Err(err).Msg("error while setting up flushing response")
+				errs.HTTPError(w, err)
+				return
+			}
+
+			// Start the event stream
+			w.Header().Set("Content-Type", "text/event-stream")
+			w.Header().Set("Cache-Control", "no-cache")
+			w.Header().Set("Connection", "keep-alive")
+			w.WriteHeader(http.StatusOK)
+			flusher.Flush()
+
+			// Run the subscription function in a goroutine
+			response := make(chan error)
```

Comment:
It's easy for cases like these to result in a leaking goroutine. Better to do:
```suggestion
			response := make(chan error, 1)
```

Reply:
I don't understand your point here, this change could result in the defer deadlocking if the context got cancelled and then the callback paniced

Reply:
What do you mean?

Reply:
By leaving the channel as unbuffered, if the parent goroutine doesn't read the value then the goroutine will block forever trying to send on `response`, which leaks a goroutine